src/stream/rabbit/rabbit.c

```diff
@@ -0,0 +1,488 @@
+
+   LTC_ARGCHK(rs != NULL);
+   LTC_ARGCHK(keystream != NULL);
+
+   /* initialize then decrement bytes_needed as bytes sent */
+   bytes_needed = length;
+
+   /*** consume any leftovers (send fm unused portion of ks) ***/
+   if (rs->unused != 0) {
+      if (length <= rs->unused) {
+         XMEMCPY(keystream, &(rs->ks[16 - rs->unused]), length);
+         bytes_sent = length;
+      } else {
+         XMEMCPY(keystream, &(rs->ks[16 - rs->unused]), rs->unused);
+         bytes_sent = rs->unused;
+      }
+      rs->unused -= bytes_sent;
+   }
+   bytes_needed -= bytes_sent;
+
+   /*** Generate and send full blocks (handle any remaining bytes later below) ***/
+   if (bytes_needed > 0) {
+       /* number of blocks to generate (incl a blk for any bytes beyond even blks) */
+       ulong32 blocks_needed = (bytes_needed +15) >> 4;  /* >>4 is quicker than /16 */
+
+       /* point out to keystream for all but last block */
+       out = keystream + bytes_sent;
+
+       /* TODO: bench this ptr for actual gains in performance */
+       /* pre-resolve this multiple-appearing constant */
+       ptr = (ulong32*)&(rs->work_ctx.x);
+       bytes_sent = 0;
+       for (i = blocks_needed; i >= 1; i--) {
+          union {
+             unsigned char  b8[16];
+             ulong32 b32[4];
+          } xover;
+
+          /* if last blk, point out to rs->ks */
+          if (i == 1) out = (unsigned char*)&(rs->ks);
+
+          /* Iterate the system */
+          _rabbit_next_state(&(rs->work_ctx));
+
+          /* Generate 16 bytes of pseudo-random data */
+          xover.b32[0] = U32TO32_LITTLE(ptr[0] ^ (ptr[5]>>16) ^ (ulong32)(ptr[3]<<16));
+          xover.b32[1] = U32TO32_LITTLE(ptr[2] ^ (ptr[7]>>16) ^ (ulong32)(ptr[5]<<16));
+          xover.b32[2] = U32TO32_LITTLE(ptr[4] ^ (ptr[1]>>16) ^ (ulong32)(ptr[7]<<16));
+          xover.b32[3] = U32TO32_LITTLE(ptr[6] ^ (ptr[3]>>16) ^ (ulong32)(ptr[1]<<16));
+          XMEMCPY(out, xover.b8, 16);
+
+          /* Increment pointers and decrement length */
+          out += 16;
+          if (i != 1) {
+             bytes_sent += 16;
+          }
+       }
+       rs->unused = 16;
+       bytes_needed -= bytes_sent;
+   }        /* end of if (bytes_needed > 0) */
+
+   /*** handle need for any remainder here ***/
+   if (bytes_needed != 0) {
+      XMEMCPY(&(keystream[length - bytes_needed]), rs->ks, bytes_needed);
+      rs->unused = (16 - bytes_needed) % 16;  /* not in inner loop so no &0xf,  */
+   }
+   return CRYPT_OK;
+}
+
+/* ------------------------------------------------------------------------- */
+
+/* Crypt a chunk of any size (encrypt/decrypt) */
+int rabbit_crypt(rabbit_state* rs, const unsigned char* input, ulong32 inlen, unsigned char* output)
+{
+   unsigned char* buffer;
+   ulong32 i;
+
+   LTC_ARGCHK(rs != NULL);
+   LTC_ARGCHK(input != NULL);
+   LTC_ARGCHK(output != NULL);
+
+   /* get right-sized keystream */
+   buffer = XMALLOC(inlen);
+   rabbit_keystream(rs, buffer, inlen);
```

Comment:
1. this ignores the return value which it most likely shouldn't
2. this is the only stream cipher that uses the pattern that `crypt()` calls `keystream()` and not vice-versa, I'd prefer to have it everywhere the same